drivers/entropy/entropy_cc13xx_cc26xx.c

```diff
@@ -83,20 +105,73 @@ static void entropy_cc13xx_cc26xx_isr(void *arg)
 	 * prevent further locking on to the sampling clock frequncy.
 	 */
 	if (src & TRNG_FRO_SHUTDOWN) {
-		/* Clear shutdown */
-		TRNGIntClear(TRNG_FRO_SHUTDOWN);
-		/* Disabled FROs */
-		off = sys_read32(TRNG_BASE + TRNG_O_ALARMSTOP);
-		/* Clear alarms */
-		sys_write32(0, TRNG_BASE + TRNG_O_ALARMMASK);
-		sys_write32(0, TRNG_BASE + TRNG_O_ALARMSTOP);
-		/* De-tune FROs */
-		sys_write32(off, TRNG_BASE + TRNG_O_FRODETUNE);
-		/* Re-enable FROs */
-		sys_write32(off, TRNG_BASE + TRNG_O_FROEN);
+		handle_shutdown_ovf();
 	}
 }
 
+static int entropy_cc13xx_cc26xx_get_entropy_isr(struct device *dev,
+	uint8_t *buf, uint16_t len, uint32_t flags)
+{
+	struct entropy_cc13xx_cc26xx_data *data = get_dev_data(dev);
+	uint16_t cnt;
+	uint16_t read = len;
+	uint32_t src;
+	uint32_t num[2];
+	unsigned int key;
+
+	key = irq_lock();
+	cnt = ring_buf_get(&data->pool, buf, len);
+	irq_unlock(key);
+
+	if ((cnt != len) && ((flags & ENTROPY_BUSYWAIT) != 0U)) {
+		buf += cnt;
+		len -= cnt;
+
+		/* Allowed to busy-wait. We should use a polling approach */
+		while (len) {
+			key = irq_lock();
+
+			src = TRNGStatusGet();
+			if (src & TRNG_NUMBER_READY) {
+				/*
+				 * This function acknowledges the ready
+				 * status
+				 */
+				num[1] = TRNGNumberGet(TRNG_HI_WORD);
+				num[0] = TRNGNumberGet(TRNG_LOW_WORD);
+
+				ring_buf_put(&data->pool, (uint8_t *)num,
+					sizeof(num));
+			}
+
+			/*
+			 * If interrupts were enabled during busy wait, this
+			 * would allow us to pick up anything that has been put
+			 * in by the ISR as well.
+			 */
+			cnt = ring_buf_get(&data->pool, buf, len);
+
+			if (src & TRNG_FRO_SHUTDOWN) {
+				handle_shutdown_ovf();
+			}
+
+			irq_unlock(key);
+
+			if (cnt) {
+				buf += cnt;
+				len -= cnt;
+			} else {
+				k_busy_wait(US_PER_SAMPLE);
```

Comment:
As IRQs are unlocked, is it better to sleep than busy wait?

Reply:
entropy_cc13xx_cc26xx_get_entropy_isr implements the version of get_entropy that is callable from isr-context, including at boot time when interrupts and threading are not yet fully enabled. So putting in k_sleep() would not be a good idea here and may lead to crashes.